### PR TITLE
Massive commit of decoupling from mongo and addinng a PanFileDB type

### PR DESCRIPTION
**Pull request by wtgee**
* PanDB is a factory class that will load either PanMongoDB or PanFileDB
* PanFileDB merely writes the json to a flat file
	* One file per "collection" along with a `current_<collection` for current
	entries. Note this means each "current" collection is a separate file.
	* Use `bson.json_util` for serializing. Changes a few things with time
	because it _can_ store timezone aware info, which we don't use properly
	in `current_time`.
* Try to only rely on one DB fro weakref kep - WIP
* Clear both mongo and file types of their "current" entries upon POCS init
* Change `include_collection` to `store_permanently` for calls to `insert_current`
* Change `use_mongo` to `store_result` where appropriate
* Private POCS method `_check_environment` changed to public class method, i.e.
	`POCS.check_environment`
* Tests that use the `db` now parameterize it across db types. Now testing
	takes almost twice as long! (should deal with this somehow)

### CODE REVIEW

**Comment by jamessynge** [comment]
I *think* we can use Python's json module, as it supports custom encoder and decoder classes. I'm willing to help with that if you like.

**Comment by wtgee** [comment]
> I think we can use Python's json module, as it supports custom encoder and decoder classes. I'm willing to help with that if you like.

I can see some advantages to a custom encoder (as mentioned ni #206) but I don't see any immediate need when this already does what we want.

**Comment by jamessynge** [comment]
Don't you have to install mongodb to use bson.json_util?

**Comment by wtgee** [comment]
No, `bson` is a standalone module. It's installed by `pymongo` but also doesn't required `pymongo`.

But I could still see some use cases for a custom encoder/decoder, just not sure it's a priority.

**Comment by jamessynge** [comment]
FWIW, I notice that we don't use bson.json_util for much: Most of the `loads()` and `dumps()` calls are using the regular json library.

```
$ egrep '\b(bson|json_util|dumps|loads)\b' $(find . -name '*.py')
./scripts/simple_sensors_capture.py:from bson import json_util
./scripts/simple_sensors_capture.py:                        f.write(json_util.dumps(data) + '\n')
./scripts/sensors_simulator.py:import bson.json_util as json_util
./scripts/collect-grid-stats.py:        f.write(json.dumps(readings))
./peas/tests/serial_handlers/protocol_arduinosimulator.py:        s = json.dumps(self.message) + '\r\n'
./peas/tests/serial_handlers/protocol_arduinosimulator.py:            message = json.loads("""
./peas/tests/serial_handlers/protocol_arduinosimulator.py:            message = json.loads("""
./pocs/dome/bisque.py:            response_obj = json.loads(response)
./pocs/mount/bisque.py:            response_obj = json.loads(response)
./pocs/camera/sbigudrv.py:        On construction loads SBIG's shared library which must have already
./pocs/utils/messaging.py:from bson import ObjectId
./pocs/utils/messaging.py:from json import dumps
./pocs/utils/messaging.py:from json import loads
./pocs/utils/messaging.py:        msg_object = dumps(message, skipkeys=True)
./pocs/utils/messaging.py:                msg_obj = loads(msg)
./pocs/utils/logger.py:    logger_key = (profile, json.dumps(log_config, sort_keys=True))
./pocs/utils/rs232.py:        return json.loads(line)
./pocs/utils/database.py:from bson import json_util
./pocs/utils/database.py:                content = json.dumps(entries, default=json_util.default)
./pocs/utils/images/cr2.py:from json import loads
./pocs/utils/images/cr2.py:        exif = loads(subprocess.check_output(cmd_list).decode('utf-8'))
./build/lib/pocs/dome/bisque.py:            response_obj = json.loads(response)
./build/lib/pocs/mount/bisque.py:            response_obj = json.loads(response)
./build/lib/pocs/camera/sbigudrv.py:        On construction loads SBIG's shared library which must have already
./build/lib/pocs/utils/messaging.py:from bson import ObjectId
./build/lib/pocs/utils/messaging.py:from json import dumps
./build/lib/pocs/utils/messaging.py:from json import loads
./build/lib/pocs/utils/messaging.py:        msg_object = dumps(message, skipkeys=True)
./build/lib/pocs/utils/messaging.py:                msg_obj = loads(msg)
./build/lib/pocs/utils/logger.py:    logger_key = (profile, json.dumps(log_config, sort_keys=True))
./build/lib/pocs/utils/rs232.py:        return json.loads(line)
./build/lib/pocs/utils/database.py:from bson import json_util
./build/lib/pocs/utils/database.py:                content = json.dumps(entries, default=json_util.default)
./build/lib/pocs/utils/images/cr2.py:from json import loads
./build/lib/pocs/utils/images/cr2.py:        exif = loads(subprocess.check_output(cmd_list).decode('utf-8'))
```

**Comment by wtgee** [comment]
I get the feeling someone doesn't want us to use `bson`. :)

We should just have a common encode/decode method in our utils folder that would let us change out the underlying implementation however we want. As for the files listed above, it's not a great example of the usage and the main ones in messaging and database tare the ones that have given us problems precisely because `json` doesn't handle things out of the box.

I'm not necessarily tied to `bson`, it just seems like we will be writing in a lot of the same functionality.

For now I'll make a common function that can be called so ti's easier to change in future.

**Comment by jamessynge** [comment]
I just don't want to be tied to Mongo. I think I now better understand that loads(default=bson.json_utils.default) provides the hook for serializing. We can leave that as is. BUT, let's put the json code into a single py file, so that we can swap it out easily.

File: pocs/tests/test_pocs.py [hunk 1]
```diff
@@ -13,8 +13,14 @@
...
 
 @pytest.fixture(scope='function')
-def observatory(config):
-    observatory = Observatory(config=config, simulator=['all'], ignore_local_config=True)
+def observatory(config, db_type):
+    observatory = Observatory(
+        config=config,
+        simulator=['all'],
+        ignore_local_config=True,
+        db_name='panoptes_testing',
```
Comment (jamessynge) [inline pocs/tests/test_pocs.py]: I think config provides this db name.

File: pocs/tests/test_pocs.py [hunk 2]
```diff
@@ -25,6 +31,7 @@ def pocs(config, observatory):
     pocs = POCS(observatory,
                 run_once=True,
                 config=config,
+                db_name='panoptes_testing',
```
Comment (jamessynge) [inline pocs/tests/test_pocs.py]: Ditto.

File: pocs/tests/test_pocs.py [hunk 3]
```diff
@@ -311,6 +323,43 @@ def test_run_complete(pocs):
 
     pocs.run(exit_when_done=True, run_once=True)
     assert pocs.state == 'sleeping'
+    pocs.power_down()
+
+
+def test_run_interrupt_with_reschedule_of_target(observatory):
```
Comment (jamessynge) [inline pocs/tests/test_pocs.py]: Probably very worthwhile, but is this related?

File: pocs/core.py [hunk 4]
```diff
@@ -366,15 +364,16 @@ def is_weather_safe(self, stale=180):
             record = self.db.get_current('weather')
 
             is_safe = record['data'].get('safe', False)
-            timestamp = record['date']
+            timestamp = record['date'].replace(tzinfo=None)
```
[the remark this answers is not in the document]
Comment (wtgee) [reply pocs/core.py]: Ah, I thought I had made a comment about this. One thing about `bson` is that it is correctly serializing the datetime and that includes timezone information. However we are not properly using that timezone information in `current_time`. Basic math between a naive datetime and a timezone-aware datetime cannot be done so here we are just dropping the tzinfo, which reverts to same behaviour as we currently have.

I had a change to `current_time(datetime=True)` that made it timezone aware based off what timezone is listed in the config file, but that had implications beyond the scope of this PR. Will file an issue about it.

File: pocs/tests/test_messaging.py [hunk 5]
```diff
@@ -68,11 +68,10 @@ def test_send_datetime(forwarder, sub, pub):
 
 def test_mongo_objectid(forwarder, sub, pub, config, db):
 
-    db.insert_current('config', {'foo': 'bar'})
+    id0 = db.insert_current('config', {'foo': 'bar'}, store_permanently=False)
```
[the remark this answers is not in the document]
Comment (wtgee) [reply pocs/tests/test_messaging.py]: It was for a mongo specific ObjectId. With flat files that is a uuid4 id. Will rename test, but fundamentally it is testing whether that is delivered via the messaging system correctly.

File: pocs/tests/test_pocs.py (hunk 3 above)
Comment (wtgee) [reply pocs/tests/test_pocs.py]: Hm, not sure why this is in this PR. 😕

**Comment by jamessynge** [comment]
I skimmed the latest 5 commits, which look reasonable and pass travis. Is this still WIP?

**Comment by wtgee** [comment]
I'm not too sure about the test_pocs.py as it now runs almost double the tests so I was going to look into changing that. I removed a few of the `sleep` calls.

But yes, I think ready for checking and I'd like to get on PAN008 soon since I can't have mongo.

**Comment by jamessynge** [comment]
No mongo because it is a Raspberry Pi?

File: pocs/tests/test_panmongo.py [hunk 6]
```diff
@@ -5,43 +5,27 @@ def test_insert_and_get_current(db):
...
-    assert record.count() == 0
-
-    db.current.remove({'type': 'config'})
 
-
-def test_insert_and_no_collection(db):
+def test_insert_and_no_permanent(db):
     rec = {'test': 'insert'}
-    db.insert_current('config', rec, include_collection=False)
+    id0 = db.insert_current('config', rec, store_permanently=False)
```
[the remark this answers is not in the document]
Comment (wtgee) [reply pocs/tests/test_panmongo.py]: It replaces `include_collection` and basically means that it stores it in the permanent collection and not just in the `current` collection.
